### ISSUE TITLE
German hyphenation: Quotation mark is speparated from related word

### ISSUE THREAD

**Issue by grabi678** [koreader/koreader#6717]
* KOReader version: v2020.09-24
* Device: Phone with Anroid 10 and PBInkPad3

#### Issue
In a book with 'unusual' german quotation maks the quotaion mark is separted from its related word by line breaks.

![grafik](https://user-images.githubusercontent.com/58222328/94345018-78794180-0023-11eb-870c-ea2fb2808960.png)

**Comment by poire-z** [comment]
German typography/line-breaking rather than German hyphenation.

Well, as you said: "unusual" german quotes - and the publisher has not set no-break-space around to prevent us from breaking there.
We break around quotes according to the default for a language, and for german, these are inverted vs most other languages. Alas, no real way to change that (that's hardcoded in the rendering code for German).

Except if you change the language to some other one that expect these quotes to be opening and closing as they are used here.
So, if you're not using hyphenation (no hyphenated word on your screenshot), you could switch `Typography language>` to `English`.
Or edit the book to switch the quotes.
Just curious: these "unusual/inverted" quotes in German text, for a German reader, are they bothering or not at all?

**Comment by hius07** [comment]
Sorry, I can't see, what is unusual here? These are usual quotation marks for russian (it is needed a period after the closing mark though).

**Comment by hius07** [comment]
Well, nevermind, germans qm are actually inverted comparing to russians. How diverse is the world!

**Comment by grabi678** [comment]
> Except if you change the language to some other one that expect these quotes to be opening and closing as they are used here.
> So, if you're not using hyphenation (no hyphenated word on your screenshot), you could switch `Typography language>` to `English`.

Switching typograhpy to English corrects displaying qm but is leading to a very fancy hyphenation.

> Just curious: these "unusual/inverted" quotes in German text, for a German reader, are they bothering or not at all?

Yes, ideed, for a German reader these qm look a bit strange but nevertheless used in German books by German publishers.
Wikipedia says that German-speaking Switzerland usually uses these qm without space:

"Switzerland  - Standard qm: | «…» | ‹…› | Alternative qm: | „…“ or “…” | ‚…‘ or ‘…’|"

Assuming that hyphenation rules for printed German in Switzerland are nearly the same as for German in Germany, isn't it possible to add these qm to German typography to also cover this case? Or adding Swiss typography and hyphenation rules?

**Comment by poire-z** [comment]
Can you look what the language of the book is in `Book information` ? and highlight a bit of text (any 2 words), `View HTML`, and see if in the HTML you see there is any `lang=` among what you see.
Is it "de" or "de-CH" or anything else ?

> Assuming that hyphenation rules for printed German in Switzerland are nearly the same as for German in Germany, isn't it possible to add these qm to German typography to also cover this case?

If we cover this case, we can't cover the regular german quotes. They are mutually exclusive :) Or it's just not covering anything and you might see bad line breaks in regular german when there is a space on both side of a `«` - these rules helps deciding what to do in this case.

> Or adding Swiss typography and hyphenation rules?

That's a better idea ! Use german hyph, but without regular german quote rules.
Needs to see what to set on:
```
de-AT : German (Austria) ??
de-CH : German (Switzerland) : no rules (or french rules?)
de-DE : German (Germany) (could be ignored = generic German)
de : German (generic) : German quotes rules as most german books must have "de" standalone
```

@roshavagarga @zwim @Frenzie : any thoughts?

**Comment by roshavagarga** [comment]
@poire-z Swiss German uses French quotation marks as their preferred, while the German style is used as an alternative as can be seen [here](https://en.wikipedia.org/wiki/Quotation_mark#Summary_table). I'm guessing the lang of that file is de-CH and we don't have quotation marks nor hyphenation patterns for that. We could get a hyphenation pattern from [here](https://github.com/hyphenation/tex-hyphen/blob/master/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/hyph-de-ch-1901.tex) if it's really required, but @zwim would have to confirm whether that's really needed, while the quotation marks should be fixable with a quick `textlang.cpp` change adding `de-ch` as long as that's the language ISO/code used in Swiss German books? We'd still need more info on what is more prevalent in modern Swiss German ebooks.

**Comment by grabi678** [comment]
> Can you look what the language of the book is in Book information ? 

Language: de

> and highlight a bit of text (any 2 words), View HTML, and see if in the HTML you see there is any lang= among what you see.
Is it "de" or "de-CH" or anything else ?

lang="de"

> Or it's just not covering anything and you might see bad line breaks in regular german when there is a space on both side of a « 

As far as I know quotation marks in German are never set with spaces before and after the included text and I have never seen this in any publication, maybe in really old books. So it couldn't happen that there is a space on both sides of a qm.

**Comment by zwim** [comment]
> @poire-z Swiss German uses French quotation marks as their preferred, while the German style is used as an alternative as can be seen [here](https://en.wikipedia.org/wiki/Quotation_mark#Summary_table). I'm guessing the lang of that file is de-CH and we don't have quotation marks nor hyphenation patterns for that. We could get a hyphenation pattern from [here](https://github.com/hyphenation/tex-hyphen/blob/master/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/hyph-de-ch-1901.tex) if it's really required, but @zwim would have to confirm whether that's really needed, while the quotation marks should be fixable with a quick `textlang.cpp` change adding `de-ch` as long as that's the language ISO/code used in Swiss German books? We'd still need more info on what is more prevalent in modern Swiss German ebooks.

In German, we have three languages: de-DE, de-AT and de-CH. One big difference is the used vocabularies. (This does not bother KOReader.)
While de-DE and de-AT are very similar (in terms of hyphenation and punctuation they are the same) except some really seldom corner cases, which we should not cover in KOReader. This has to be done by the book authors/publishing company!
de-CH is a bit different. The Swiss don't use "ß" at all, instead of that they use "ss". Therefore, different hyph-patterns have to be used. IIRC punctuation differs only for Guillemots.

In de-DE and de-AT we use: »Hallo« or ›Hallo‹ (no spaces)
in de-CH it is: «Hallo» or ‹Hallo› (no spaces)

Don't use the patterns from https://github.com/hyphenation/tex-hyphen/blob/master/hyph-utf8/tex/generic/hyph-utf8/patterns/tex/hyph-de-ch-1901.tex as this is the traditional=very old writing. Since we (DE,AT,CH) had spelling reforms 2000, 2004 and 2006. I can provide actual patterns for de-CH from http://projekte.dante.de/Trennmuster if this is wanted.

**In summary:** An actual pattern file would be needed for de-CH and if the iso-code is de-CH, KOReader should not do linebrake after `‹`, `«` and before `›`, `»`

**Comment by zwim** [comment]
> As far as I know quotation marks in German are never set with spaces before and after the included text and I have never seen this in any publication, maybe in really old books. So it couldn't happen that there is a space on both sides of a qm.

That is right. No space on **both** sides of a punctuation mark, except (emdash, endash, - and ellipsis).

**Comment by poire-z** [comment]
> No space on **both** sides of a punctuation mark

Both ? :) That's not what we see in the screenshot above. On one side I would believe you :)

> KOReader should not do linebrake after ‹, « and before ›, »

That's what it ensures for French.
And the reverse of generic German.
But not for most other languages.

I had been explained how that works at https://github.com/koreader/crengine/issues/364#issuecomment-671079147:
> QU allows no breaks unless there are spaces. That is indeed a difference with OP and CL, which forbid breaks (on one side) even with spaces

So, for french (resp. german), we set OP(ening) and CL(osing) on « and » (resp » and «).
On most other languages, these are set QU(ote) = we can't know if they are opening or closing - or the "forbid break even with space" rule is not needed/to be avoided.

In French, that's necessary as we can have "j'ai dit « bonjour » et je suis parti".

> As far as I know quotation marks in German are never set with spaces before and after the included text and I have never seen this in any publication, maybe in really old books. 

So, not as necessary for German? Or may be in some other conditions? libunibreak (the library we use) sets OP/CL for german, so I guess it's needed for something.
May be you can imagine a case where it feels useful from my quotation above?
Anyway, it doesn't feel we should remove that OP/CL for »/« for German.

But I'm fine with adding de-CH / German (Switzerland or Swiss?). With the french quotes rules (or no rules so it's available as a fallback for mixed quote texts...)?
Even with the classic German hyphenation. It shouldn't be that different, is it?

> The Swiss don't use "ß" at all, instead of that they use "ss". Therefore, different hyph-patterns have to be used. IIRC punctuation differs only for Guillemots.

Would this be an issue if the same pattern that have "ß" would be duplicated with "ss" in main the German.pattern ? Can they be occurences of "ss" in german German ?

**Comment by zwim** [comment]
@poire-z: First I want to state that the hyphenation and the hanging punctuation for de-AT and de-DE works very good. I have read quite a few books now and I have not found any problems (except some wrong patterns, which were corrected by upstream by dante.de/Trennmuster). So for de-AT, de-DE there is no need for changes.

> So, not as necessary for German? Or may be in some other conditions? libunibreak (the library we use) sets OP/CL for german, so I guess it's needed for something.
> May be you can imagine a case where it feels useful from my quotation above?

It could be possible that someone uses `»` as a relational operator in a formula (like `v « c`), then there would be a space before and after, or maybe someone uses `«` as bullet-points in an enumeration. There could be more cases.

On the initial commit of this issue: The second marked `«` in `« Meine Vermutung` is syntactically wrong. It should be `«Meine Vermutung` (without a space). I assume the there is also a space after `»` in the second line, which causes the linebreak.

> Anyway, it doesn't feel we should remove that OP/CL for »/« for German.

I agree totally.

> Would this be an issue if the same pattern that have "ß" would be duplicated with "ss" in main the German.pattern ? Can they be occurences of "ss" in german German ?

The idea sounds good at the first thought. But the Trennmuster-Team at DANTE takes care of all three German variants.  I think, they know more about right hyphenation than I do, so I would not mess with replacing `ß` with `ss`. 

I have generated the de-CH patterns. At a quick glance there are 200 patterns more (for major hyphenation points) in de-CH as in de-AT, de-DE. 
If we will use two different pattern files for German, wouldn't it be good to name them German_DE and German_CH?



> But I'm fine with adding de-CH / German (Switzerland or Swiss?). With the french quotes rules (or no rules so it's available as a fallback for mixed quote texts...)?

I agree with you. For the naming I would prefer de-CH, as Schweizerdeutsch is just a variant of German and not a sole language. Don`t we use en-US and en-GB, too?

**Comment by poire-z** [comment]
> It could be possible that someone uses » as a relational operator in a formula (like v « c), then there would be a space before and after, or maybe someone uses « as bullet-points in an enumeration. There could be more cases.

I doubt it :) It they do, they deseve to have crappy rendering.

> On the initial commit of this issue: The second marked « in « Meine Vermutung is syntactically wrong. It should be «Meine Vermutung (without a space). I assume the there is also a space after » in the second line, which causes the linebreak.

I think your german nature made you get that one wrong :) It does not mark the start of `»Meine` but the end of `«Was ... Wikinger?»` followed by `Meine` :) It feels like this text should be unreadable by a german person (you'd have to go back to look what started where...)

So, it's probably because of these hard rules that there is a line break _even if there is not space after « in the second line_ (unless we're not talking about the same thing).

But I dunno. https://en.wikipedia.org/wiki/Quotation_mark#German says:
> This German double-quote style is also used in the Netherlands, but is falling out of fashion nowadays[when?] with the 'English-style' quotation marks being preferred. However, it still can be found on older shop signs and in most large newspapers. 

So, if it were to happen that they are becoming old fashioned (and there's always no space on the good side), and that more and more "English-style" guillemets might be seen in German text/books, may be we should revisit this:
https://github.com/koreader/crengine/blob/a00d1b8c26f4a621d805575a6a4cf10d81c8e7c7/crengine/src/textlang.cpp#L705-L713
You might try to comment out:
```
        has_left_double_angle_quotation_mark_closing = true;
        has_right_double_angle_quotation_mark_opening = true;
```
(and may be others?), make a build and see if you notice some linebreaking glitches in your regular reading (I guess this should fix the issue of the first post). If you notice some glitch, we might better understand why these are needed.

> If we will use two different pattern files for German, wouldn't it be good to name them German_DE and German_CH?

Best to keep `German.pattern` for the regular German (to avoid changes, duplicate files on user devices, and the need for tweaks in the old setting migration code).

> I agree with you. For the naming I would prefer de-CH, as Schweizerdeutsch is just a variant of German and not a sole language. Don`t we use en-US and en-GB, too?

We'll need an english translation for "Schweizerdeutsch" :) (but that's for frontend).

It would require this kind of changes (order matters: `de-CH` must be matched before `de`):
```diff
--- a/crengine/src/textlang.cpp
+++ b/crengine/src/textlang.cpp
@@ -49,2 +49,3 @@ static struct {
     { "ka",    "Georgian",      "Georgian.pattern",      1, 2 },
+    { "de-CH", "German_CH",     "German_CH.pattern",     2, 2 },
     { "de",    "German",        "German.pattern",        2, 2 },
@@ -297,2 +298,3 @@ static quotes_spec _quotes_spec_table[] = {
     { "da",       L"\x201c", L"\x201d", L"\x2018", L"\x2019" }, /* â<80><9C> â<80><9D> â<80><98> â<80><99> */
+    { "de-CH",    "with the right stuff" },
     { "de",       L"\x201e", L"\x201c", L"\x201a", L"\x2018" }, /* â<80><9E> â<80><9C> â<80><9A> â<80><98> */
@@ -694,3 +696,3 @@ TextLangCfg::TextLangCfg( lString16 lang_tag ) {
     }
-    else if ( LANG_STARTS_WITH(("fr") ("es")) ) { // French, Spanish
+    else if ( LANG_STARTS_WITH(("fr") ("es") ("de-CH") ) { // French, Spanish, Schweizerdeutsch
         has_left_single_quotation_mark_opening = true; // no right..closing in linebreakdef.c
@@ -704,2 +706,5 @@ TextLangCfg::TextLangCfg( lString16 lang_tag ) {
     }
+    else if ( LANG_STARTS_WITH(("de-CH")) ) { // or this if no need for hard French rules
+        // Skip "de" rules
+    }
     else if ( LANG_STARTS_WITH(("de")) ) { // German
```

**Comment by Frenzie** [comment]
> This German double-quote style is also used in the Netherlands, but is falling out of fashion nowadays[when?] with the 'English-style' quotation marks being preferred. However, it still can be found on older shop signs and in most large newspapers.

It's wrong. We quote „Like this.” Not „Like that (what it calls German style).“

Or “Like this.”

**Comment by zwim** [comment]
>I think your german nature made you get that one wrong :) It does not mark the start of »Meine but the end of «Was ... Wikinger?» followed by Meine :) It feels like this text should be unreadable by a german person (you'd have to go back to look what started where...)

You are right, I am not used to the Swiss quotations as I am an Austrian.  But now I got it: Is there a space in the book between the first marked « and Reinickendorf? There should not be one (do we know that?),  and there shouldn't be a line break either.
Between the second marked » and the ? there should also be no space (and no linebreak).

> but is falling out of fashion nowadays[when?] with the 'English-style' quotation marks being preferred.
That is not quite right. In handwritten texts only the

**Comment by grabi678** [comment]
> You are right, I am not used to the Swiss quotations as I am an Austrian. But now I got it: Is there a space in the book between the first marked « and Reinickendorf? There should not be one (do we know that?), and there shouldn't be a line break either.
Between the second marked » and the ? there should also be no space (and no linebreak).

Above the same text with smaller font size. In the whole book there are no spaces after opening and before closing quotation marks.

![grafik](https://user-images.githubusercontent.com/58222328/94468008-0f7cff80-01c4-11eb-8efa-8d179e09212e.png)

Another example: If a qm stands at the very end of a paragraph and of a full line it is solely wrapped to a new line. 

![grafik](https://user-images.githubusercontent.com/58222328/94468957-8666c800-01c5-11eb-8002-96683ba5b52e.png)

**Comment by zwim** [comment]
@poire-z: I have done
```
--- a/crengine/src/textlang.cpp
+++ b/crengine/src/textlang.cpp
@@ -49,2 +49,3 @@ static struct {
     { "ka",    "Georgian",      "Georgian.pattern",      1, 2 },
+    { "de-CH", "German_CH",     "German_CH.pattern",     2, 2 },
     { "de",    "German",        "German.pattern",        2, 2 },
@@ -297,2 +298,3 @@ static quotes_spec _quotes_spec_table[] = {
     { "da",       L"\x201c", L"\x201d", L"\x2018", L"\x2019" }, /* â<80><9C> â<80><9D> â<80><98> â<80><99> */
+        { "de-CH",  L"\x201e", L"\x201c", L"\x201a", L"\x2018" }, /* „ “ ‚ ‘ */ /*the same as in de*/

     { "de",       L"\x201e", L"\x201c", L"\x201a", L"\x2018" }, /* â<80><9E> â<80><9C> â<80><9A> â<80><98> */
@@ -694,3 +696,3 @@ TextLangCfg::TextLangCfg( lString16 lang_tag ) {
     }
-    else if ( LANG_STARTS_WITH(("fr") ("es")) ) { // French, Spanish
+    else if ( LANG_STARTS_WITH(("fr") ("es") ("de-CH") ) { // French, Spanish, Schweizerdeutsch
         has_left_single_quotation_mark_opening = true; // no right..closing in linebreakdef.c
@@ -704,2 +706,5 @@ TextLangCfg::TextLangCfg( lString16 lang_tag ) {
     }
+    else if ( LANG_STARTS_WITH(("de-CH")) ) { // or this if no need for hard French rules
+        // Skip "de" rules
+    }
     else if ( LANG_STARTS_WITH(("de")) ) { // German
```
and 
```
    else if ( LANG_STARTS_WITH(("de")) ) { // German
        has_left_single_quotation_mark_closing = true;
        has_right_single_quotation_mark_glue = true;
        has_left_double_quotation_mark_closing = true;
//        has_left_single_angle_quotation_mark_closing = true;
//        has_right_single_angle_quotation_mark_opening = true;
//        has_left_double_angle_quotation_mark_closing = true;
//        has_right_double_angle_quotation_mark_opening = true;
    }
```
and I had no negative effects on the books I have checked.

**Comment by poire-z** [comment]
OK, so for "de", commenting out the 4 angle quote marks one should be fine and solve the top post issue.
Can you make sense of the 3 first rules : left single & double quotation mark being closing ? and right sing quotation mark glue ?

The, for de-CH, I don't know if it should have the rigid French rules (nearly mirror of what was there for "de" and that you softened). Would they use space around angle quotes in de-CH like we do in french ? Or do they stick it to words and the rules might not be needed (and thus, it would allows normal-german-inverted-quotes found in de-CH text to work quite OK (instead of having the same kind of issue as the top post, but inverted).

**Comment by zwim** [comment]
> OK, so for "de", commenting out the 4 angle quote marks one should be fine and solve the top post issue.
> Can you make sense of the 3 first rules : left single & double quotation mark being closing ? and right sing quotation mark glue ?

Sorry, I don't have a clue what they mean. I could comment them out and see what is happening.
I ping you again, when I had time to test this with some books.

> 
> The, for de-CH, I don't know if it should have the rigid French rules (nearly mirror of what was there for "de" and that you softened). Would they use space around angle quotes in de-CH like we do in french ? Or do they stick it to words and the rules might not be needed (and thus, it would allows normal-german-inverted-quotes found in de-CH text to work quite OK (instead of having the same kind of issue as the top post, but inverted).

As I see the only difference between de-CH and the other de-??s is, that the Guillemets are the other way round. All other rules arte the same. So the Guillemets stick to the word.

 In the whole de- family no punctuation mark stands free except the dashes and the ellipsis.

**Comment by roshavagarga** [comment]
@poire-z The Wikipedia article has them sticking to words, so as long as it's correct there...

**Comment by poire-z** [comment]
```
        has_left_single_quotation_mark_closing = true;
        has_left_double_quotation_mark_closing = true;
```
means like we had for angle quotations - but here only for the left ones :/ 
So, these rules make `U+2018 ‘` and `U+201C “` considered closing: if there is a space before, the will stick with the space and the word before the space (and so, I guess, more easily allow a break on their right side).

`        has_right_single_quotation_mark_glue = true;`
This makes `U+2019 ’` sticking I guess :), like a no-break space, preventing any break on both sides (?)
Knowing the German language, does that have some sense?

It's a bit obscure why all these rules (no comment about why they were set in libunibreak). I can guess why those for french are, but that's it.

English has `has_left_single_quotation_mark_opening = true; ` - but nothing for the right_single_quotation_mark_closing, may be because `’` might happen in more english constructs like `isn't it` or `good o' man` or something :)

**Comment by poire-z** [comment]
I commented out the 4 has_left/right_**single/double_angle_quotation_mark**_opening/closing for de as part of https://github.com/koreader/crengine/pull/385.
We'll then see if/why/how much these are needed :)

**Comment by poire-z** [comment]
Issue closed, but keep discussing de-CH (make a PR in crengine if it's worth it) and the effect of my above change.
